migrator.py

```diff
@@ -3,39 +3,153 @@
+            global_names[contact.attrib['id']] = contact.attrib['name']
+
+    else:
+        # Alternate way using *.ini files
+        # Sometimes names will be missing or conflicting if found 
+        # different in multiple directories.  Conflicting names
+        # will be merged by this script by concatenating them with " | " 
+        # separators in sorted order.  Missing names will have an
+        # auto generated value of  
+        contact_id_2_nameset: Dict[str, set] = {}
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+                learn_contact_ids(input_dir,ini_file,contact_id_2_nameset)
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+                learn_old_contact_ids(input_dir,ini_file,contact_id_2_nameset)
+
+        logging.info("contact_id_2_nameset")
+        logging.info("------------------------------------------------------------")
+        try:
+            for contact_id, names in contact_id_2_nameset.items():
+                person_name = ""
+                for i, item in enumerate(sorted(names)):
+                    person_name += item
+                    if i < (len(names)-1):
+                        # ambiguous contact_id will have a concatenated list of names
+                        person_name += "|"
+                logging.info(f"'{contact_id}': '{person_name}'")
+                # Create or add contacts
+                global_names[contact_id] = person_name
+        except Exception as e:
+            logging.warning(f"Exception: {e}")
+        logging.info("------------------------------------------------------------")
+
     contact_tags_per_dir: Dict[Path, ContactTags] = {}
-    for input_dir, subdirs, files in os.walk(input_root_dir):
-        dir = Path(input_dir)
-        if _PICASA_INI_FILE in files:
-           contact_tags_per_dir[dir] = migrate_directory(dir, files, db, contact_tags_per_dir, dry_run=dry_run)
-        elif any([_is_photo_file(file) for file in files]):
-            logging.warning('Found photos but no %s in %s' % (_PICASA_INI_FILE, dir))
 
+    for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            dir = Path(input_dir)
+            logging.info(f"Processing {Path(dir/ini_file)}")
+            try:
+                if ini_file in files:
+                    contact_tags_per_dir[dir] = migrate_directory(dir, files, db, 
+                                                    contact_tags_per_dir, global_names,
+                                                    dry_run=dry_run, ini_file_name=ini_file)
+                elif any([_is_photo_file(file) for file in files]):
+                    logging.warning(f'Found photos but no {ini_file} in {dir}')
+            except Exception as e:
```

Comment:
Why is catching necessary here?

Reply:
Happens when digiKam scan skips certain directories.  The Windows version skips my "Originals" directories.  Weird thing is Linux version doesn't.